internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// depending on the start and end time we may get len(results) <=  buckets + 1
+	// pass the max number of values to getHitsFromHistoryResponse to ensure we only get the desired number of values
+	return getHitsFromHistoryResponse(body, params.Buckets), nil
+}
+
+// setDefaultHistoryParams sets the default values for the history params.
+func setDefaultHistoryParams(params *GetHistoryParams) {
+	if params.Buckets == 0 {
+		params.Buckets = 1000
+	}
+	if params.EndTime.IsZero() {
+		params.EndTime = time.Now()
+	}
+
+	if params.StartTime.IsZero() {
+		params.StartTime = params.EndTime.Add(-time.Hour * 24 * 14) // default to 2 weeks ago
+	}
+}
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateInterval(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
+
+	msm := types.MinimumShouldMatch(1)
+	min := 1
+	filter := getHistoricalPrivilegeFilter(params.PrivilegeIDs)
+	return search.Request{
+		// get documents that match the device ID, are within the time range, and have a vehicle field
+		Query: &types.Query{
+			Bool: &types.BoolQuery{
+				Filter: []types.Query{
+					{Term: map[string]types.TermQuery{subjectField: {Value: params.DeviceID}}},
+					{Range: map[string]types.RangeQuery{timeField: types.DateRangeQuery{Gte: &startArg, Lte: &endArg}}},
+				},
+				Should: []types.Query{
+					{Exists: &types.ExistsQuery{Field: "data.vehicle"}},
+				},
+				MinimumShouldMatch: &msm,
+			},
+		},
+		Aggregations: map[string]types.Aggregations{
+			// docByInterval histogram aggregates the data into buckets based on the interval
+			docByInterval: {
+				DateHistogram: &types.DateHistogramAggregation{
+					Field:         ptr(timeField),
+					FixedInterval: &intervalArg,
+					MinDocCount:   &min,
+				},
+				Aggregations: map[string]types.Aggregations{
+					// singleDoc aggregation selects the first data point in each bucket
+					// that matches the desired filter
+					singleDoc: {
+						TopHits: &types.TopHitsAggregation{
+							Size:    ptr(1),
+							Source_: &filter,
+						},
+					},
+				},
+			},
+		},
+		Size: ptr(0), // only return the aggregations
+	}
+}
+
+// calculateInterval calculates the desired interval in milliseconds. This interval will provide n buckets between the start and end times.
+// if the interval is 0 or negative it will be clamped to 1ms.
+func calculateInterval(start, end time.Time, buckets int) int {
+	duration := int(end.Sub(start).Milliseconds())
```

Comment:
Do we check that `start` and `end` are "in order" anywhere?

Reply:
Note to myself: `int` is 64 bits on 64 bit systems, so this cast is okay.

Reply:
No, but we do test that this does not break anything. If this is a negative number, the interval will be negative and be clamped to 1. Elastic will return an empty list since no time will match your time range.

Reply:
> Do we check that `start` and `end` are "in order" anywhere?

Ah, there's a negative check.

Reply:
This is one of a few places where it's kinda hard to follow. It's okay though, there are tests.